File: mobile/ios/Runner/FaceCheckPlugin.swift
import Flutter
import UIKit
import Vision

/// Answers the Dart side's `skidoo/face_check` channel: "does this photo
/// contain a face?"
///
/// Vision ships with iOS, so this costs nothing in the app download. It
/// replaced ML Kit's bundled face detector, which linked ~16 MB of static code
/// into the binary and shipped a further 10 MB of TFLite models for the same
/// yes/no.
///
/// Returning `nil` means "couldn't tell" — the Dart side treats that as a pass
/// rather than blocking the user (see `FaceCheck.hasFace`).
enum FaceCheckPlugin {
  static let channelName = "skidoo/face_check"

  static func register(with registry: FlutterPluginRegistry) {
    // Any registrar's messenger reaches the same engine; this plugin has no
    // platform views or assets, so the name is only an identity.
    let registrar = registry.registrar(forPlugin: "FaceCheckPlugin")!
    let channel = FlutterMethodChannel(
      name: channelName,
      binaryMessenger: registrar.messenger()
    )

    channel.setMethodCallHandler { call, result in
      guard call.method == "hasFace" else {
        result(FlutterMethodNotImplemented)
        return
      }
      guard
        let args = call.arguments as? [String: Any],
        let path = args["path"] as? String
      else {
        result(nil)
        return
      }

      // Decoding and detection are both slow enough to matter on the platform
      // thread; the reply hops back to main because the Flutter messenger
      // expects it there.
      DispatchQueue.global(qos: .userInitiated).async {
        let answer = detectFace(atPath: path)
        DispatchQueue.main.async { result(answer) }
      }
    }
  }

  private static func detectFace(atPath path: String) -> Bool? {
    guard
      let image = UIImage(contentsOfFile: path),
      let cgImage = image.cgImage
    else { return nil }

    let request = VNDetectFaceRectanglesRequest()
    let handler = VNImageRequestHandler(
      cgImage: cgImage,
      // A selfie is usually stored rotated, with the upright direction only in
      // the EXIF orientation. Vision detects on the raw buffer, so without
      // this a portrait capture arrives sideways and finds no face.
      orientation: cgOrientation(of: image.imageOrientation),
      options: [:]
    )

    do {
      try handler.perform([request])
      return !(request.results ?? []).isEmpty
    } catch {
      return nil
    }
  }

  private static func cgOrientation(
    of orientation: UIImage.Orientation
  ) -> CGImagePropertyOrientation {
    switch orientation {
    case .up: return .up
    case .down: return .down
    case .left: return .left
    case .right: return .right
    case .upMirrored: return .upMirrored
    case .downMirrored: return .downMirrored
    case .leftMirrored: return .leftMirrored
    case .rightMirrored: return .rightMirrored
    @unknown default: return .up
    }
  }
}

File: mobile/android/app/src/main/kotlin/com/skidoo/app/FaceCheckPlugin.kt
package com.jperg.app

import android.content.Context
import android.net.Uri
import com.google.mlkit.vision.common.InputImage
import com.google.mlkit.vision.face.FaceDetection
import com.google.mlkit.vision.face.FaceDetectorOptions
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import java.io.File

/**
 * Answers the Dart side's `jperg/face_check` channel: "does this photo
 * contain a face?"
 *
 * Uses the *unbundled* ML Kit face detector — the model lives in Play Services
 * and is downloaded on demand rather than shipped in the APK. The bundled
 * `google_mlkit_face_detection` plugin this replaced carried the same detector
 * and its models inside the app, for ~16 MB.
 *
 * Replies `null` for "couldn't tell" — no model yet, an unreadable file, a
 * detector error. The Dart side treats that as a pass rather than blocking the
 * user (see `FaceCheck.hasFace`).
 */
class FaceCheckPlugin(private val context: Context) : MethodChannel.MethodCallHandler {

    companion object {
        const val CHANNEL = "jperg/face_check"

        fun register(context: Context, messenger: BinaryMessenger) {
            MethodChannel(messenger, CHANNEL).setMethodCallHandler(FaceCheckPlugin(context))
        }
    }

    private val options = FaceDetectorOptions.Builder()
        .setPerformanceMode(FaceDetectorOptions.PERFORMANCE_MODE_FAST)
        .setMinFaceSize(0.15f)
        .build()

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        if (call.method != "hasFace") {
            result.notImplemented()
            return
        }

        val path = call.argument<String>("path")
        if (path.isNullOrEmpty() || !File(path).exists()) {
            result.success(null)
            return
        }

        val image = try {
            InputImage.fromFilePath(context, Uri.fromFile(File(path)))
        } catch (e: Exception) {
            result.success(null)
            return
        }

        val detector = FaceDetection.getClient(options)
        detector.process(image)
            .addOnSuccessListener { faces ->
                result.success(faces.isNotEmpty())
                detector.close()
            }
            .addOnFailureListener {
                // Most often the model is still downloading on a fresh install.
                result.success(null)
                detector.close()
            }
    }
}
